DSpace/dspace-replicate
# Package Role Information in BagIt Exports

Resolves: https://jira.lyrasis.org/browse/DS-4481

* Adds Pojos for tags defined in the DSpaceRoles schema
* Add a utility class for retrieving and ingesting DSpaceRoles from DSpaceObjects
* Package DSpaceRoles as roles.xml for BagIt exports
* Read DSpaceRoles during import and ingest using the RoleIngester

--------------------------------

## Notes

The work for retrieving roles for DSpaceObjects was taken heavily from the [RoleDisseminator](https://github.com/DSpace/DSpace/blob/master/dspace-api/src/main/java/org/dspace/content/packager/RoleDisseminator.java) and updated to fit in a way which works with how xml writing is handled for BagIt exports. Since the xml schema for the DSpaceRoles is shared between implementations, the [RoleIngester](https://github.com/DSpace/DSpace/blob/master/dspace-api/src/main/java/org/dspace/content/packager/RoleIngester.java) is used in order to provide ingestion. 

There is a difference in that BagIt implementation will not emit passwords for any `EPerson` as there isn't a `passwords` option in the RTS configuration. If needed this can be added on to the current implementation.

-----------------------------------

## Testing

1. Create a new DSpace repository with the RTS and bagit enabled
   a. In your `config/dspace.cfg`, add the following lines
    ```
    include = ${module_dir}/replicate.cfg
    include = ${module_dir}/replicate-bagit.cfg
    ```
    b. Copy all configuration files from dspace-replicate to your DSpace installation module config, e.g. `config/modules/`
    c. Check that bagit packaging is enabled in your `config/modules/replicate.cfg`
    ```
    replicate.packer.pkgtype = bagit
    replicate.packer.archfmt = tar
    ```
    d. Check that `config/modules/replicate-bagit.cfg` has set all required fields for the BTR profle
    ```
    replicate-bagit.profile = beyondtherepository
    replicate-bagit.tag.bag-info.source-organization = org.dspace
    ```
    e. Check that all rts tasks are setup in `config/modules/curate.cfg`
    ```
    plugin.named.org.dspace.curate.CurationTask = \
    org.dspace.ctask.replicate.EstimateAIPSize = estaipsize, \
    org.dspace.ctask.replicate.ReadOdometer = readodometer, \
    org.dspace.ctask.replicate.TransmitAIP = transmitaip, \
    org.dspace.ctask.replicate.TransmitSingleAIP = transmitsingleaip, \
    org.dspace.ctask.replicate.VerifyAIP = verifyaip, \
    org.dspace.ctask.replicate.FetchAIP = fetchaip, \
    org.dspace.ctask.replicate.CompareWithAIP = auditaip, \
    org.dspace.ctask.replicate.RemoveAIP = removeaip, \
    org.dspace.ctask.replicate.BagItRestoreFromAIP = restorefromaip, \
    org.dspace.ctask.replicate.BagItReplaceWithAIP = replacewithaip
    ```
2. Create a Community, Collection, and Item
3. (Optional) Edit the authorizations for each of the above
    a. From the Admin Panel, `Access Control > Authorizations`
    b. If needed, a group can be created through `Access Control > Group` in order to test managed groups
4. Assign roles for the Community
    a. Select you Community, then from the context panel choose `Edit Community`
    b. Select `Assign Roles` and create an Administrator role
5. Assign roles for the Collection
    a. Select you Collection, then from the context panel choose `Edit Collection`
    b. Select `Assign Roles` and create any desired roles for the Collection
6. Run the `Transfer AIP(s)` curation option for the community
7. Review the `data/roles.xml` written for...
    a. the Community
    b. the Collection
8. Delete some of the roles from the `Assign Roles` menu for the Community and Collection
9. Run the `Replace all AIPs` curation option for the Community
10. Review the roles and verify they match what exist when the initial transfer was done
